lib/rules/no-redeclare.js

```diff
@@ -84,6 +86,129 @@ module.exports = {
 			}
 		}
 
+		/**
+		 * Determines if a definition node is of a specific type.
+		 * @param {ASTNode} node The AST node to check.
+		 * @param {string} nodeType The node type to check against.
+		 * @returns {boolean} True if the node is of the specified type.
+		 */
+		function isNodeType(node, nodeType) {
+			return node && node.type === nodeType;
+		}
+
+		/**
+		 * Checks if a combination of TypeScript declarations is valid.
+		 * Valid combinations:
+		 * - multiple interfaces (interface merging)
+		 * - interface + class
+		 * - class + namespace
+		 * - interface + class + namespace
+		 * - function + namespace
+		 * - multiple namespaces
+		 * @param {Array} nodes Array of declaration nodes.
+		 * @returns {boolean} True if the combination is valid in TypeScript.
+		 */
+		function isValidTSDeclarationCombination(nodes) {
+			if (nodes.length < 2) {
+				return false;
+			}
+
+			const counts = {
+				interfaces: nodes.filter(node =>
+					isNodeType(node, "TSInterfaceDeclaration"),
+				).length,
+				classes: nodes.filter(node =>
+					isNodeType(node, "ClassDeclaration"),
+				).length,
+				namespaces: nodes.filter(node =>
+					isNodeType(node, "TSModuleDeclaration"),
+				).length,
+				functions: nodes.filter(node =>
+					isNodeType(node, "FunctionDeclaration"),
+				).length,
+				others: nodes.filter(
+					node =>
+						!isNodeType(node, "TSInterfaceDeclaration") &&
+						!isNodeType(node, "ClassDeclaration") &&
+						!isNodeType(node, "TSModuleDeclaration") &&
+						!isNodeType(node, "FunctionDeclaration"),
+				).length,
+			};
+
+			// If there are non-TS declarations mixed in, not valid
+			if (counts.others > 0) {
+				return false;
+			}
+
+			// Check valid combinations
+			const { interfaces, classes, namespaces, functions } = counts;
+
+			// Multiple interfaces only (interface merging)
+			if (
+				interfaces > 1 &&
+				classes === 0 &&
+				namespaces === 0 &&
+				functions === 0
+			) {
+				return true;
+			}
+
+			// Multiple namespaces only
+			if (
+				namespaces > 1 &&
+				interfaces === 0 &&
+				classes === 0 &&
+				functions === 0
+			) {
+				return true;
+			}
+
+			// Interface + class (+ optional namespace)
+			if (interfaces === 1 && classes === 1 && functions === 0) {
+				return true;
+			}
+
+			// Class + namespace (no interface or function)
+			if (
+				classes === 1 &&
+				namespaces >= 1 &&
+				interfaces === 0 &&
+				functions === 0
+			) {
+				return true;
+			}
+
+			// Function + namespace (no interface or class)
+			if (
+				functions === 1 &&
+				namespaces >= 1 &&
+				interfaces === 0 &&
+				classes === 0
+			) {
+				return true;
+			}
```

Comment:
With this logic, the rule is making explicit exceptions for certain combinations of declaration types that should not be reported.

I think it would be easier to do the opposite: start by allowing all combinations, and report an error only for specific cases, for example when `counts.others > 0`, `counts.functions > 1`, and probably not much more.